### ISSUE TITLE
Very slow 

### ISSUE THREAD

**Issue by Kalelv45** [xtr33me/Reinforcement_Learning_for_Stock_Prediction#3]
There is any particolar reason of why this is running so slow. I tried to remove many date from the data samples (leaving only the days where BTC was above 1000$), yet to do 1000 ep will take me about 3 days.

**Comment by smileyung** [comment]
Having the same feeling,
I use the GTX1080Ti and train with slow speed.

**Comment by Bluesn0w** [comment]
delete the following ones from the agent should make it much faster. You can also move the first one to function. I am not quite sure about the 2nd one, but there is other ways to calculate learning rate. 


def setupTensorboardDir(self):
		current_dir = os.path.dirname(os.path.realpath(__file__))
		self.log_dir = os.path.join(current_dir, '../tensorboard')
		print('Working directory: %s' % current_dir)
		if not os.path.exists(self.log_dir) or not os.path.isdir(self.log_dir):
			os.mkdir(self.log_dir)

class LRTensorBoard(TensorBoard):
	def __init__(self, *args, **kwargs):
		#self.scalar = kwargs.pop('scalar', True)
		super(LRTensorBoard, self).__init__(*args, **kwargs)

		global tf
		import tensorflow as tf

	def on_epoch_end(self, epoch, logs=None):
		logs.update({'lr': K.eval(self.model.optimizer.lr)})
		super(LRTensorBoard, self).on_epoch_end(epoch, logs)

**Comment by xtr33me** [comment]
Thanks @Bluesn0w for the find.  I was attempting to add logic as you can see for Tensorboard graphing. Ofcourse for this to work in Keras, one has to use their callback functionality. I haven't used it before in Keras and this was the solution to integrate TB when using Keras. Have to look into whether there is a better way of doing this.  That said, glad that your find seemed to have resolved your speed issues.  Just out of curiosity, what was the speed slow down you saw with the logic?  I am currently training another project on my GPU so not able to test...just curious.

**Comment by Bluesn0w** [comment]
My pleasure! I didn't measure how much faster, but it is definitely much faster. Also in the exreplay function, change the first 4 lines of codes to the following can speed it up even more.

	def expReplay(self, batch_size):
		mini_batch = []
		l = len(self.memory)
		for i in range(l - batch_size + 1, l):
			mini_batch.append(self.memory.popleft())

However, i didn't find it is that accurate, so I will spend more time over the weekend to see if anything else can be improved. Please keep us posted if you have more findings too.

**Comment by smileyung** [comment]
@Bluesn0w 
amazing!
why this way could improve the speed?

**Comment by xtr33me** [comment]
Truly incredible the speed difference @Bluesn0w . Thanks again!  I have made the advised changes and removed the Tensorboard handling for now and committed. I will see if there are more advisable ways to use TB when using keras.  Going to close this now.
